### DISCUSSION TITLE
Enable Link-Time Optimization (LTO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [brocode/fblog#106] [Ideas]
Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) for the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance a bit.

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where additionally to regular `release` optimizations LTO will also be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO should also help to reduce the build-time overhead with LTO. If we enable it on the Cargo profile level, users, who install the application with `cargo install`, will get the LTO-optimized version "automatically". E.g., check `cargo-outdated` Release [profile](https://github.com/kbknapp/cargo-outdated/blob/master/Cargo.toml#L48).

Basically, it can be enabled with the following lines:

```
[profile.release]
lto = true
```

Thank you.

**Comment by icepuma** [comment]
https://github.com/johnthagen/min-sized-rust is also a good source on how to do this :)

**Reply by zamazan4ik** [reply]
Yeah! I also can recommend a handy Cargo extension - https://github.com/Kobzol/cargo-wizard

**Comment by bomgar** [comment]
I had to deactivate it [here](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=fblog#n9) explicitly. If I enable it there it breaks the build.

I did not investigate this further. But I don't want to enable it if I don't understand why this breaks.

**Reply by zamazan4ik** [reply]
> If I enable it there it breaks the build.

Do you have at least the build logs from the broken build? It would be pretty interesting to find the root cause of the problem.

**Reply by bomgar** [reply]
The log is longer than that but a lot of those:
```
  = note: /usr/bin/ld: fblog-aur/src/fblog-4.13.0/target/release/deps/fblog-50db4c743190de0e.fblog.e5a3c9bd97f2765e-cgu.02.rcgu.o: in function `mlua::function::Function::call':
          fblog.e5a3c9bd97f2765e-cgu.02:(.text._ZN4mlua8function8Function4call17h23ab9af08fb82300E+0x23): undefined reference to `lua_gettop'
          /usr/bin/ld: fblog.e5a3c9bd97f2765e-cgu.02:(.text._ZN4mlua8function8Function4call17h23ab9af08fb82300E+0x39): undefined reference to `lua_checkstack'
          /usr/bin/ld: fblog.e5a3c9bd97f2765e-cgu.02:(.text._ZN4mlua8function8Function4call17h23ab9af08fb82300E+0x54): undefined reference to `lua_pushcclosure'
          /usr/bin/ld: fblog.e5a3c9bd97f2765e-cgu.02:(.text._ZN4mlua8function8Function4call17h23ab9af08fb82300E+0x5d): undefined reference to `lua_gettop'
          /usr/bin/ld: fblog.e5a3c9bd97f2765e-cgu.02:(.text._ZN4mlua8function8Function4call17h23ab9af08fb82300E+0x83): undefined reference to `lua_pcallk'
          /usr/bin/ld: fblog.e5a3c9bd97f2765e-cgu.02:(.text._ZN4mlua8function8Function4call17h23ab9af08fb82300E+0xd3): undefined reference to `lua_gettop'
          /usr/bin/ld: fblog.e5a3c9bd97f2765e-cgu.02:(.text._ZN4mlua8function8Function4call17h23ab9af08fb82300E+0xea): undefined reference to `lua_toboolean'
```

**Reply by zamazan4ik** [reply]
Does it come from `lto = true` (aka Fat LTO)? Did you try to use `lto = "thin"` (aka Thin LTO) instead?

Seems like one of the project's dependencies is surprisingly not okay with enabling LTO.

**Reply by bomgar** [reply]
I did not modify the Cargo.toml. It seems to be enabled by default with pkgbuild

**Reply by zamazan4ik** [reply]
What kind of LTO is enabled by pkgbuild by default in Arch? Is it possible to override this setting if we want for example use Thin LTO instead of Fat?

**Reply by bomgar** [reply]
I don't know. I did not have the time to investigate further.
